### ISSUE TITLE
Large Frequent Library Updates 

### ISSUE THREAD

**Issue by brando56894** [jellyfin/jellyfin-kodi#509]
**Describe the bug**
Upon launching Kodi, after Jellyfin logs in, there are a large amount of updates (hundreds) occurring when maybe only a few new files were added. 

Please see here for my detailed report and other user reports: https://www.reddit.com/r/jellyfin/comments/ms4ufq/jellyfin_for_kodi_always_has_a_huge_number_of/

**To Reproduce**
Not exactly sure, but I believe it's related to the TMBD Box Sets plugin. I had it happening almost daily, then while waiting for @mcarlton00  to build the debug plugin, I found that apparently TMDB Box Sets has an issue where if you add something to a collection and then change the name/location of the file the plugin keeps looking for it at the old location ([bug report](https://github.com/jellyfin/jellyfin/issues/1907)). I attempted to fix that just by removing the TMDB Box Sets plugin but the "bad" records were still in the database. I deleted just my movies library and rescanned them in, without the Box Sets plugin enabled. I had zero or very few updates during the next few days. I reinstalled it and rescanned the collections, over the next few days things were being updated, and the large batches of updates started again. It was a lot lower than before, only a hundred or so, but then yesterday it popped up for 700+ notifications **when only one new item had been added that day**.

**Expected behavior**
There are no large daily updates when nothing has changed.

**Logs**
I can't post the text anywhere because it's about 186,000 lines for about 4 hours worth of logs, the file is 17 MB, so I hosted it on my Google Drive: https://drive.google.com/file/d/1rrmqmvW55GenbDq9WXu3itIKoMLf2nTv/view?usp=sharing

**System (please complete the following information):**
 - OS: Arch Linux (Jellyfin Server)/Nvidia Shield Android TV v8.2.2 (Kodi client)
 - Jellyfin Version: 10.7.2
 - Kodi Version: 19.0.0
 - Addon Version: 0.7.1+py3
 - Playback Mode: Native

**Comment by mcarlton00** [comment]
This seems to confirm my theory that the server is giving us unnecessary updates, but some thoughts on your logs.

--------------------------------

First off, debug logging in this case doesn't particularly help, unfortunately.  It gives us the full metadata for each object, but since we don't know what that metadata was before these most recent changes it doesn't give us a whole lot of insight to what's happened.  That can significantly cut down on your log size if we need to collect more.

------------------------------

Second, it looks like 27 of the updates you received were "valid".  Maybe not necessarily wanted, given how they changed, but the changes we received from the server do explain why Kodi updated them.  You can see these with
```
% grep -A 2 column kodi.log | sed -En "s/(.*)movies.py:6[0-9]//p"
```
A lot of them seem to be getting switched from having a value to None.  I can't imagine why that's happened, but it makes sense why we'd be getting change requests for that.

---------------------------------------

Now the more interesting part, and what seems to indicate my theory is correct, is the number of updates we received that had no changes in the database.

```
% grep 'no changes' kodi.log | wc -l
754
```

Looks to be a lot of American Dad, Family Guy, and Whose Line in the list.  You can see episode titles with

```
% grep 'no changes' kodi.log | sed -En "s/^(.*).py:[0-9]+ //p"
```

So this seems to mean is that the kodi sync queue plugin is marking things as updated even if we don't necessarily care about the piece of data that's been updated.  We were discussing this in chat the other day and started some thoughts on it here https://github.com/jellyfin/jellyfin-plugin-kodisyncqueue/issues/44

---------------------------------------

Overall, it sounds like there's nothing we can do about it _here_ and I'll need to go beg the server/plugin folks to poke around there, but it's good reference material to influence future API decisions.  The bit about TMDB Boxsets having an effect here is very odd, but good information to have and another avenue to look into.

**Comment by brando56894** [comment]
Interesting! Glad you were able to glean some information from that. I wasn't sure if debug logs were going to be helpful since you modded the plugin yourself to get more info, I'll disable it now in case it happens again. 

I'm going to mass rename the files/folders my movie library to make them all homogenous and then I can see if that causes huge daily updates again, because I haven't really had any since I posted this.

**Comment by razordullwit** [comment]
I'm encountering a similar issue. Almost every time I open Kodi on my Shield TV, it wants to sync my entire library, even if I've only updated a few things. Expected behavior would be that it only syncs new or actually changed items.

The problem is definitely upstream, but I can't figure out _what_ is triggering the updates. I can't figure out if it's a scheduled task or a plugin or what. It's not logged anywhere, and I don't know how to access the database (nor would I necessarily want to).

Has anyone found the primary culprit for problems like this yet? Googling has not helped.

**Comment by elgatopanzon** [comment]
I just connected a new setup of Kodi 21 to my existing Jellyfin server, and installed the Kodi Sync Queue plugin. On my jellyfin server I have a scheduled task to update my library every 30 mins. Unfortunately, this pushes a 14k item update queue to Kodi every 30 minutes. It takes around 20-25 mins to finish that update, so Kodi is in a constant state of updating even when nothing changed in the jellyfin library.

**Comment by oddstr13** [comment]
The APIs unfortunately don't tell us what changes, just that an item has changed – we'd likely need to mirror the whole media library in the sync queue plugin in order to filter what is of interest to the Kodi plugin or not.

By the looks of it, any change to the BaseItem object likely triggers an update, maybe even the `DateLastSaved` and `DateLastRefreshed` fields...
https://github.com/jellyfin/jellyfin/blob/a7bb3ea21489a3a4c9a3e9f60c9521eaa84bad31/MediaBrowser.Controller/Entities/BaseItem.cs#L422

To get any significant improvement here, we'd need changes in the Jellyfin server, as we are only listening to the "item has changed" event.
https://github.com/jellyfin/jellyfin-plugin-kodisyncqueue/blob/089fecd966d00929ce80b2e6af0adb1ac720a3c9/Jellyfin.Plugin.KodiSyncQueue/EntryPoints/LibrarySyncNotification.cs#L75

**Comment by razordullwit** [comment]
> The APIs unfortunately don't tell us what changes, just that an item has changed – we'd likely need to mirror the whole media library in the sync queue plugin in order to filter what is of interest to the Kodi plugin or not.
> 

Is there any way to tell what's triggering this? Because nothing HAS changed, but every single library item gets flagged with the has changed flag per the log. I can't tell what causes it, as far as I can tell it's not any plug-in that's doing it.

**Comment by bdzerr** [comment]
Yeah this is also happening to me and gets rather annoying, loading kodi up daily and having thousands of updates pending everytime regardless if I have actually added 1 item or not, ive tried everything I can think of, disabling plugins, even tried disabling metadata updates completely and yet still thousands of updates pending, im a stump

**Comment by oddstr13** [comment]
You could enable debug logging in jf4kodi, and manually compare the json objects of a single item getting updated, or modify the syncqueue plugin to do the same directly in the mentioned event, comparing a saved baseitem to the new one (with the matching item id)

**Comment by elgatopanzon** [comment]
I was able to figure out what was causing this for me. I have a library in jellyfin for Youtube archived videos. I checked, and there's 14k "episodes" in that library. I also confirmed this with the debug log in the kodi addon which revealed this:

`2025-05-08 22:31:31.463 T:9108     info <general>: JELLYFIN.jellyfin_kodi.helper.utils -> ERROR::jellyfin_kodi\helper\utils.py:583 No ancestor found, not syncing item with ID: 8bca03d0ac18869c781d7b6a0927d8dc`

I grepped the log for `No ancestor found`, and there were 14k matches which is around the amount of processing updates I get each time, and essentially matched the episode count for my Youtube library. The IDs (I checked a bunch) point to items in that library.

A couple of things about this:
1. This library is NOT in the jellyfin for kodi whitelist
2. Oddly enough the library in jellyfin reported 2 copies for every video, despite there only being 1 file
3. Syncing this library to kodi did NOT fix the issue

In the end, I deleted the whole library in jellyfin and then removed it from the kodi side, and the issue mostly resolved itself. Doing a full library scan still gives me a few items to process but it's much less, it's around 400 items. It finishes it in about 10 seconds. Unlike before, there are no log errors for those items.

I'll try re-adding the youtube library and seeing if that fixes the double-video bug, and maybe that stops the entire 14k videos showing up as not having an ancestor.
